tools/copyblocks/main.go

```diff
@@ -0,0 +1,507 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/concurrency"
+	"github.com/grafana/dskit/flagext"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"google.golang.org/api/iterator"
+
+	"github.com/grafana/mimir/pkg/storage/tsdb/bucketindex"
+)
+
+const (
+	delim = "/" // Used by Mimir to delimit tenants and blocks, and objects within blocks.
+)
+
+type config struct {
+	sourceBucket      string
+	destBucket        string
+	minBlockDuration  time.Duration
+	minTime           flagext.Time
+	maxTime           flagext.Time
+	tenantConcurrency int
+	blocksConcurrency int
+	copyPeriod        time.Duration
+	enabledUsers      flagext.StringSliceCSV
+	disabledUsers     flagext.StringSliceCSV
+	dryRun            bool
+
+	httpListen string
+}
+
+func (c *config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&c.sourceBucket, "source-bucket", "", "Source GCS bucket with blocks.")
+	f.StringVar(&c.destBucket, "destination-bucket", "", "Destination GCS bucket with blocks.")
+	f.DurationVar(&c.minBlockDuration, "min-block-duration", 0, "If non-zero, ignore blocks that cover block range smaller than this.")
+	f.Var(&c.minTime, "min-time", fmt.Sprintf("If set, only blocks with MinTime >= this value are copied. The supported time format is %q.", time.RFC3339))
+	f.Var(&c.maxTime, "max-time", fmt.Sprintf("If set, only blocks with MaxTime <= this value are copied. The supported time format is %q.", time.RFC3339))
+	f.IntVar(&c.tenantConcurrency, "tenant-concurrency", 5, "How many tenants to process at once.")
+	f.IntVar(&c.blocksConcurrency, "block-concurrency", 5, "How many blocks to copy at once per tenant.")
+	f.DurationVar(&c.copyPeriod, "copy-period", 0, "How often to repeat the copy. If set to 0, copy is done once, and program stops. Otherwise program keeps running and copying blocks until terminated.")
```

Comment:
```suggestion
	f.DurationVar(&c.copyPeriod, "copy-period", 0, "How often to repeat the copy. If set to 0, copy is done once, and the program stops. Otherwise, the program keeps running and copying blocks until it is terminated.")
```